mdformat/_util.py

```
@@ -1,21 +1,81 @@
+        plugin.update_mdit(mdit)
+        ignore_classes.extend(getattr(plugin, "ignore_classes", []))
+    html1 = HTML2JSON().parse(mdit.render(md1), ignore_classes)
+    html2 = HTML2JSON().parse(mdit.render(md2), ignore_classes)
+
     return html1 == html2
+
+
+class HTML2JSON(HTMLParser):
+    """Parser HTML to JSON."""
+
+    def parse(self, text: str, strip_classes: Iterable[str] = ()) -> List[dict]:
+        self.tree: List[dict] = []
+        self.current: Optional[dict] = None
+        self.feed(text)
+        self.strip_classes(self.tree, set(strip_classes))
+        return self.tree
+
+    def strip_classes(self, tree: List[dict], classes: Set[str]) -> List[dict]:
+        """Strip content from tags with certain classes."""
+        items = []
+        for item in tree:
+            if set(item["attrs"].get("class", "").split()).intersection(classes):
+                items.append({"tag": item["tag"], "attrs": item["attrs"]})
+                continue
+            items.append(item)
+            item["children"] = self.strip_classes(item.get("children", []), classes)
+            if not item["children"]:
+                item.pop("children")
+
+        return items
+
+    def handle_starttag(self, tag: str, attrs: list) -> None:
+        tag_item = {"tag": tag, "attrs": dict(attrs), "parent": self.current}
+        if self.current is None:
+            self.tree.append(tag_item)
+        else:
+            children = self.current.setdefault("children", [])
+            children.append(tag_item)
+        self.current = tag_item
+
+    def handle_endtag(self, tag: str) -> None:
+        # walk up the tree to the tag's parent
+        while self.current is not None:
+            if self.current["tag"] == tag:
+                self.current = self.current.pop("parent")
+                break
+            self.current = self.current.pop("parent")
+
+    def handle_data(self, data: str) -> None:
+        if self.current:
```

Can `self.current` ever be falsy, given valid HTML? If not, maybe remove the check and fail fast and hard?

---

well self.current can technically be `None` (although it will have been replaced in `handle_starttag` by this point) and so mypy complained about it

---

Aah if its just mypy complaining, I'd help mypy with a `assert self.current is not None` instead of the `if` clause